addons/l10n_it_edi/models/account_move.py

```diff
@@ -0,0 +1,1212 @@
+                    tax_dict['base_amount'] = base_amount - tax_dict['rounding']
+
+            tax_line_dict = {
+                'tax': tax,
+                'rounding': tax_dict.get('rounding', False),
+                'base_amount': tax_dict['base_amount'],
+                'tax_amount': tax_dict['tax_amount'],
+            }
+            tax_lines.append(tax_line_dict)
+        return tax_lines
+
+    def _l10n_it_edi_filter_tax_details(self, line, tax_values):
+        """Filters tax details to only include the positive amounted lines regarding VAT taxes."""
+        repartition_line = tax_values['tax_repartition_line']
+        return (repartition_line.factor_percent >= 0 and repartition_line.tax_id.amount >= 0)
+
+    def _l10n_it_edi_get_values(self, pdf_values=None):
+        self.ensure_one()
+
+        # Flags
+        is_self_invoice = self._l10n_it_edi_is_self_invoice()
+        document_type = self._l10n_it_edi_get_document_type()
+
+        # Represent if the document is a reverse charge refund in a single variable
+        reverse_charge = document_type in ['TD17', 'TD18', 'TD19']
+        is_downpayment = document_type in ['TD02']
+        reverse_charge_refund = self.move_type == 'in_refund' and reverse_charge
+        convert_to_euros = self.currency_id.name != 'EUR'
+
+        tax_details = self._prepare_invoice_aggregated_taxes(filter_tax_values_to_apply=self._l10n_it_edi_filter_tax_details)
+
+        company = self.company_id
+        partner = self.commercial_partner_id
+        sender = company
+        buyer = partner if not is_self_invoice else company
+        seller = company if not is_self_invoice else partner
+        sender_info_values = company.partner_id._l10n_it_edi_get_values()
+        buyer_info_values = (partner if not is_self_invoice else company.partner_id)._l10n_it_edi_get_values()
+        seller_info_values = (company.partner_id if not is_self_invoice else partner)._l10n_it_edi_get_values()
+        representative_info_values = company.l10n_it_tax_representative_partner_id._l10n_it_edi_get_values()
+
+        if self._l10n_it_edi_is_simplified_document_type(document_type):
+            formato_trasmissione = "FSM10"
+        elif partner._is_pa():
+            formato_trasmissione = "FPA12"
+        else:
+            formato_trasmissione = "FPR12"
+
+        # Self-invoices are technically -100%/+100% repartitioned
+        # but functionally need to be exported as 100%
+        document_total = self.amount_total
+        if is_self_invoice:
+            document_total += sum([abs(v['tax_amount_currency']) for k, v in tax_details['tax_details'].items()])
+            if reverse_charge_refund:
+                document_total = -abs(document_total)
+
+        # Reference line for finding the conversion rate used in the document
+        conversion_line = self.invoice_line_ids.sorted(lambda l: abs(l.balance), reverse=True)[0] if self.invoice_line_ids else None
```

Comment:
Wtf is that?
The rate of the invoice is `abs(invoice.amount_total / invoice.amount_total_signed)`

Reply:
Sure, I can change it to `abs(invoice.amount_total / invoice.amount_total_signed)` but none of these methods is able to completely eliminate the error caused by the rounding of the AMLs.

If we wanted to find an optimal solution, we could consider this problem from a signal processing perspective, seeing that each invoice line has an amount_currency equal to balance * rate + noise where the noise is introduced by rounding, and then find an optimal estimator for the rate. Note that the optimal estimator would give less weighting to lines with a lower balance since they would have a lower signal-to-noise ratio, so using `abs(invoice.amount_total / invoice.amount_total_signed)` (which gives an equal weighting to all the lines) is not exactly optimal either. 

Anyway we will soon have the currency rate directly on the invoice, so won't need to worry about this anymore...
https://github.com/odoo/odoo/pull/133299